File: 3383-taking-maximum-energy-from-the-mystic-dungeon/3383-taking-maximum-energy-from-the-mystic-dungeon.py
class Solution:
    def maximumEnergy(self, energy: List[int], k: int) -> int:
        n = len(energy)
        dp = [0] * n
        ans = float('-inf')

        for i in range(n - 1, -1, -1):
            dp[i] = energy[i]
            if i + k < n:
                dp[i] += dp[i + k]
            ans = max(ans, dp[i])
        return ans

File: 3383-taking-maximum-energy-from-the-mystic-dungeon/3383-taking-maximum-energy-from-the-mystic-dungeon.cpp
class Solution {
public:
    long long maximumEnergy(vector<int>& energy, int k) {
        int n = energy.size();
        vector<long long> dp(n, 0);
        long long ans = LLONG_MIN;

        for (int i = n - 1; i >= 0; i--) {
            dp[i] = energy[i];
            if (i + k < n) dp[i] += dp[i + k];
            ans = max(ans, dp[i]);
        }
        return ans;
    }
};
